File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int K = 0;
    int ans = 0;
    
    int kthSmallest(TreeNode* root, int k)
    {
        dfs(root, k);
        return ans;
    }
    
    void dfs(TreeNode* root, int k)
    {
        if(root == nullptr) return;
        
        dfs(root->left, k);
        
        K++;
        
        if(K == k)
        {
            ans = root->val;
        }
        dfs(root->right, k);
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        ret = [0, 0]
        def dfs(node):
            if not node:
                return
            
            dfs(node.left)
            
            ret[0] += 1
            if ret[0] == k:
                ret[1] = node.val
                return
            
            dfs(node.right)
            
        dfs(root)
        
        return ret[1]